evcouplings/align/protocol.py

```diff
@@ -38,6 +40,106 @@
+    The search result is filtered to only contain the columns with
+    match states to the HMM, which has a one to one mapping to the
+    query sequence.
+
+    Paramters
+    ---------
+    ar: dict
+        Alignment result dictionary, output by run_hmmsearch
+    prefix: str
+        Prefix for file creation
+
+    Returns
+    -------
+    ali: EVcouplings.align.Alignment
+        Alignment object
+
+    """
+    def _add_gaps_to_query(query_sequence_ali, ali):
+
+        new_sequence = ""
+        seq = list(query_sequence_ali .matrix[0, :])
+
+        # loop through every position in the HMMsearch hits
+        for i in range(0, len(ali.annotation["GC"]["RF"])):
+            # if that position should be a gap, add a gap
+            if i in gap_index:
+                new_sequence += "-"
+            # if that position should be a letter, pop the next
+            # letter in the query sequence
+            else:
+                new_sequence += seq.pop(0)
+
+        new_sequence_ali = Alignment.from_dict({
+            query_sequence_ali.ids[0]: new_sequence
+        })
+        return new_sequence_ali
+
+    # open the sequence file
+    with open(ar["target_sequence_file"]) as a:
+        query_sequence_ali = Alignment.from_file(a, format='fasta')
+
+    # if the provided alignment is empty, just return the target sequence 
+    if not valid_file(ar["raw_alignment_file"]):
+
+        # write the query sequence to a fasta file
+        with open(prefix + "_raw.fasta", "w") as of:
+            query_sequence_ali.write(of)
+
+        # return as an alignment object
+        return query_sequence_ali
+
+    # else, open the HMM search result
+    with open(ar["raw_alignment_file"]) as a:
+        ali = Alignment.from_file(a, format='stockholm')
+
+    # make sure that the stockholm alignment contains the match annotation
+    if not ("GC" in ali.annotation and "RF" in ali.annotation["GC"]):
+        raise ValueError(
+            "Stockholm alignment {} missing RF"
+            " annotation of match states".format(ar["raw_alignment_file"])
+        )
+            
+    # get the index of columns that do not contain match states (indicated by an x)
+    gap_index = [i for i, x in enumerate(ali.annotation["GC"]["RF"]) if x != "x"]
+    # get the index of columns that contain match states (indicated by an x)
+    match_index = [i for i, x in enumerate(ali.annotation["GC"]["RF"]) if x == "x"]
+
+    # ensure that the length of the match states 
+    # match the length of the sequence
+    if len(match_index) != query_sequence_ali.L:
+        raise ValueError(
+            "HMMsearch result {} does not have a one-to-one"
+            " mapping to the query sequence columns".format(ar["raw_alignment_file"])
+        )
+
+    # add insertions to the query sequence in order to preserve correct
+    # numbering of match sequences
+    gapped_sequence_ali = _add_gaps_to_query(query_sequence_ali, ali)
```

Comment:
the code determining match indices seems a bit redundant to what happens inside _add_gaps_to_query, maybe move that check inside the sub-function?